File: _skels/next-js-ddd-skel/src/categories/adapters/sql.js
/**
 * Drizzle-backed implementation of CategoryRepository.
 */

const { eq, asc } = require('drizzle-orm');
const { categories } = require('../../lib/schema');
const { wrapDb } = require('../../shared/errors');

class DrizzleCategoryRepository {
  constructor(db) {
    if (!db) throw new Error('DrizzleCategoryRepository requires a Drizzle db handle');
    this.db = db;
  }

  list() {
    return this.db.select().from(categories).orderBy(asc(categories.name)).all();
  }

  get(id) {
    return (
      this.db.select().from(categories).where(eq(categories.id, Number(id))).get() || null
    );
  }

  create({ name, description }) {
    try {
      return this.db
        .insert(categories)
        .values({ name, description: description ?? null })
        .returning()
        .get();
    } catch (err) {
      throw wrapDb(err);
    }
  }

  update(id, patch) {
    const categoryId = Number(id);
    const existing = this.get(categoryId);
    if (!existing) return null;

    const next = {
      name: patch.name !== undefined ? patch.name : existing.name,
      description:
        patch.description !== undefined ? patch.description : existing.description,
      updated_at: new Date(),
    };

    try {
      return this.db
        .update(categories)
        .set(next)
        .where(eq(categories.id, categoryId))
        .returning()
        .get();
    } catch (err) {
      throw wrapDb(err);
    }
  }

  delete(id) {
    const categoryId = Number(id);
    const existing = this.get(categoryId);
    if (!existing) return false;
    this.db.delete(categories).where(eq(categories.id, categoryId)).run();
    return true;
  }
}

module.exports = { DrizzleCategoryRepository };

File: _skels/rust-axum-ddd-skel/src/categories/adapters/sql.rs
//! SeaORM-backed implementation of `CategoryRepository`.

use std::sync::Arc;

use async_trait::async_trait;
use chrono::Utc;
use sea_orm::{
    ActiveModelTrait, DatabaseConnection, EntityTrait, ModelTrait, QueryOrder, Set,
};

use crate::categories::repository::{Category, CategoryRepository, NewCategory};
use crate::entities::category;
use crate::shared::{errors::is_unique_violation, DomainError};

#[derive(Clone)]
pub struct SeaCategoryRepository {
    db: Arc<DatabaseConnection>,
}

impl SeaCategoryRepository {
    pub fn new(db: Arc<DatabaseConnection>) -> Self {
        Self { db }
    }
}

#[async_trait]
impl CategoryRepository for SeaCategoryRepository {
    async fn list(&self) -> Result<Vec<Category>, DomainError> {
        let rows = category::Entity::find()
            .order_by_asc(category::Column::Name)
            .all(self.db.as_ref())
            .await?;
        Ok(rows)
    }

    async fn get(&self, id: i32) -> Result<Category, DomainError> {
        let row = category::Entity::find_by_id(id)
            .one(self.db.as_ref())
            .await?;
        row.ok_or_else(|| DomainError::NotFound(format!("category {id} not found")))
    }

    async fn create(&self, new: NewCategory) -> Result<Category, DomainError> {
        let now = Utc::now();
        let active = category::ActiveModel {
            name: Set(new.name.clone()),
            description: Set(new.description),
            created_at: Set(now),
            updated_at: Set(now),
            ..Default::default()
        };
        match active.insert(self.db.as_ref()).await {
            Ok(model) => Ok(model),
            Err(err) => {
                if is_unique_violation(&err) {
                    Err(DomainError::Conflict(format!(
                        "category '{}' already exists",
                        new.name
                    )))
                } else {
                    Err(DomainError::Db(err))
                }
            }
        }
    }

    async fn update(&self, id: i32, new: NewCategory) -> Result<Category, DomainError> {
        let row = category::Entity::find_by_id(id)
            .one(self.db.as_ref())
            .await?;
        let existing =
            row.ok_or_else(|| DomainError::NotFound(format!("category {id} not found")))?;
        let mut active: category::ActiveModel = existing.into();
        active.name = Set(new.name.clone());
        active.description = Set(new.description);
        active.updated_at = Set(Utc::now());
        match active.update(self.db.as_ref()).await {
            Ok(model) => Ok(model),
            Err(err) => {
                if is_unique_violation(&err) {
                    Err(DomainError::Conflict(format!(
                        "category '{}' already exists",
                        new.name
                    )))
                } else {
                    Err(DomainError::Db(err))
                }
            }
        }
    }

    async fn delete(&self, id: i32) -> Result<(), DomainError> {
        let row = category::Entity::find_by_id(id)
            .one(self.db.as_ref())
            .await?;
        let existing =
            row.ok_or_else(|| DomainError::NotFound(format!("category {id} not found")))?;
        existing.delete(self.db.as_ref()).await?;
        Ok(())
    }
}
